network/src/peer/fuzzing.rs

```diff
@@ -0,0 +1,150 @@
+    },
+    testutils::fake_socket::ReadOnlyTestSocketVec,
+    transport::{Connection, ConnectionId, ConnectionMetadata},
+    ProtocolId,
+};
+use futures::{executor::block_on, future, io::AsyncReadExt, sink::SinkExt, stream::StreamExt};
+use libra_config::network_id::NetworkContext;
+use libra_logger::prelude::*;
+use libra_network_address::NetworkAddress;
+use libra_proptest_helpers::ValueGenerator;
+use libra_types::PeerId;
+use memsocket::MemorySocket;
+use netcore::{compat::IoCompat, transport::ConnectionOrigin};
+use proptest::{arbitrary::any, collection::vec};
+use std::time::Duration;
+use tokio_util::codec::FramedWrite;
+
+/// Generate a sequence of `NetworkMessage`, lcs serialize them, and write them
+/// out to a buffer using our length-prefixed message codec.
+pub fn generate_corpus(gen: &mut ValueGenerator) -> Vec<u8> {
+    let network_msgs = gen.generate(vec(any::<NetworkMessage>(), 1..20));
+
+    let network_msgs = network_msgs
+        .into_iter()
+        .map(|msg| lcs::to_bytes(&msg).unwrap())
+        .collect::<Vec<_>>();
+
+    let (write_socket, mut read_socket) = MemorySocket::new_pair();
+    let mut writer = FramedWrite::new(IoCompat::new(write_socket), message_codec(MAX_FRAME_SIZE));
+
+    // Write the `NetworkMessage`s to a fake socket
+    let f_send = async move {
+        for network_msg in network_msgs {
+            writer.send(network_msg.into()).await.unwrap();
+        }
+    };
+    // Read the serialized `NetworkMessage`s from the fake socket
+    let f_recv = async move {
+        let mut buf = Vec::new();
+        read_socket.read_to_end(&mut buf).await.unwrap();
+        buf
+    };
+
+    let (_, buf) = block_on(future::join(f_send, f_recv));
+    buf
+}
+
+/// Fuzz the `Peer` actor's inbound message handling.
+///
+/// For each fuzzer iteration, we spin up a new `Peer` actor and pipe the raw
+/// fuzzer data into it. This mostly tests that the `Peer` inbound message handling
+/// doesn't panic or leak memory when reading, deserializing, and handling messages
+/// from remote peers.
+pub fn fuzz(data: &[u8]) {
+    // Use the basic single-threaded runtime, since our current tokio version has
+    // a chance to leak memory and/or thread handles when using the threaded
+    // runtime and sometimes blocks when trying to shutdown the runtime.
+    //
+    // https://github.com/tokio-rs/tokio/pull/2649
+    let mut rt = tokio::runtime::Builder::new()
+        .basic_scheduler()
+        .enable_all()
+        .build()
+        .unwrap();
+    let executor = rt.handle().clone();
+
+    // We want to choose a constant peer id for _our_ peer id, since we will
+    // generate unbounded metrics otherwise and OOM during fuzzing.
+    let peer_id = PeerId::ZERO;
+    // However, we want to choose a random _remote_ peer id to ensure we _don't_
+    // have metrics logging the remote peer id (which would eventually OOM in
+    // production for public-facing nodes).
+    let remote_peer_id = PeerId::random();
```

Comment:
I'm not sure I understand why you want a random remote_peer_id here, and I would also suggest against it since this might trip the fuzzer (we want to avoid any non-determinism)

Reply:
i'm using this as a low-tech assertion that we don't have any per-remote-peer metrics, since it'll eventually OOM if we do : p

unfortunately, we're already going to get some nasty non-determinism since we're using async stuff (the basic scheduler should limit this at least) and internal timers : (

Reply:
wait, but isn't it a legit OOM then? Remember that the fuzzer resets the state of everything between runs

Reply:
the llvm fuzzer doesn't spin up a separate process for each iter, so static/global state is preserved between iters. i don't know about the google OSS fuzzer, maybe you can comment on that.

in this specific case, we have a global metrics collector (basically a big global `HashMap<MetricName, Counter>`). if every fuzzer iter pokes a counter with a different name/set of labels this big hashmap will blow up and we'll OOM.

as noted in the comment above this comment, this is expected behaviour in some cases. for example, if we use a different self-peer-id every fuzzer iter, we'll quickly OOM since we include the self-peer-id as a label in all network metrics.

Reply:
ah, any ways to disable these globals when fuzzing :D? Where is this being hit? Like, does that impact other fuzzers as well?

Reply:
nope! no way to disable them that i know of. for metrics, the global is the default metrics registry: https://docs.rs/prometheus/0.10.0/src/prometheus/registry.rs.html#309

this is one reason why globals suck, but it probably beats passing around a `&Metrics` handle everywhere

Reply:
Damn that looks weird to me, I'd imagine a better way would be to have an initialized static + counters function only when initialized = true